pkg/cloud/gcp/actuators/util/gcp_tags_labels.go

```diff
@@ -119,3 +162,114 @@ func GetLabelsList(userLabelsAllowed bool, client controllerclient.Client, machi
 
 	return labels, nil
 }
+
+// getInfraResourceTagsList returns the user-defined tags present in the
+// status sub-resource of Infrastructure.
+func getInfraResourceTagsList(platformStatus *configv1.PlatformStatus) (tags []machinev1.ResourceManagerTag) {
+	if platformStatus != nil && platformStatus.GCP != nil && platformStatus.GCP.ResourceTags != nil {
+		tags = make([]machinev1.ResourceManagerTag, len(platformStatus.GCP.ResourceTags))
+		for i, tag := range platformStatus.GCP.ResourceTags {
+			tags[i] = machinev1.ResourceManagerTag(tag)
+		}
+	}
+	return
+}
+
+// getTagClientOptions returns the tag client options adding the credentials and
+// the endpoint which will be used by the client.
+func getTagClientOptions(cred string) []option.ClientOption {
+	return []option.ClientOption{
+		option.WithCredentialsJSON([]byte(cred)),
+		option.WithEndpoint(fmt.Sprintf("https://%s", resourceManagerHostSubPath)),
+	}
+}
+
+// getTagValuesClient returns the client to be used for reading tag values to
+// the resources.
+func (t *tagManager) getTagValuesClient(ctx context.Context) (*tagValuesClient, error) {
+	client, err := rscmgr.NewTagValuesRESTClient(ctx, t.tagClientOptions...)
+	return &tagValuesClient{client}, err
+}
+
+func (c *tagValuesClient) getTagValuesNames(ctx context.Context, tagList []machinev1.ResourceManagerTag) (map[string]string, error) {
+	tagValueList := make(map[string]string, len(tagList))
+	getTagValuesReq := &rscmgrpb.GetNamespacedTagValueRequest{}
+	for _, tag := range tagList {
+		getTagValuesReq.Name = fmt.Sprintf("%s/%s/%s", tag.ParentID, tag.Key, tag.Value)
+		value, err := c.GetNamespacedTagValue(ctx, getTagValuesReq)
+		if err != nil {
+			var gErr *apierror.APIError
+			// google API returns StatusForbidden or StatusNotFound when the tag
+			// does not exist, since it could be because of permission issues
+			// or genuinely tag does not exist.
+			if errors.As(err, &gErr) && (gErr.HTTPCode() == http.StatusNotFound ||
+				gErr.HTTPCode() == http.StatusForbidden) {
+				klog.Errorf("does not have permission to access %s tag or tag does not exist", getTagValuesReq.Name)
+				continue
+			}
+			return nil, err
+		}
+		tagValueList[value.Parent] = value.Name
+	}
+
+	return tagValueList, nil
+}
+
+// mergeInfraProviderSpecTags merges user-defined tags in Infrastructure.Status and
+// GCPMachineProviderSpec, with precedence given to those in GCPMachineProviderSpec
+// for new or updated tags. infraTags arg will contain the merged list.
+func mergeInfraProviderSpecTags(infraTags *[]machinev1.ResourceManagerTag, providerSpecTags []machinev1.ResourceManagerTag) {
+	for _, pTag := range providerSpecTags {
+		appendTag := true
+		for i, iTag := range *infraTags {
+			if pTag.ParentID == iTag.ParentID && pTag.Key == iTag.Key && pTag.Value == iTag.Value {
+				appendTag = false
```

Comment:
Just append directly?

Reply:
I have updated the logic, but `appendTag` flag is still used to skip the duplicate tags.